src/include/skiplist_template.h

```diff
@@ -0,0 +1,369 @@
+/*-
+ * Copyright (c) 2014-present MongoDB, Inc.
+ * Copyright (c) 2008-2014 WiredTiger, Inc.
+ *	All rights reserved.
+ *
+ * See the file LICENSE for redistribution information.
+ */
+
+/*
+ * PLEASE READ THE FOLLOWING DESCRIPTION CAREFULLY BEFORE INCLUDING THIS FILE.
+ *
+ * This is a template for building skip lists. Ideally, we would like to use C++ templates, but
+ * because this language feature is not available in C, we have to do use the following trick: Set
+ * up all relevant function names, type names, and "template" parameters using macros, and only then
+ * include this file. This will define the corresponding templatized functions and types.
+ *
+ * Please refer to skiplist.h for additional details and for an example of these macros.
+ */
+
+/*
+ * TODO: We would like additional functions here, such as going forward and backward within a skip
+ * list, or certain functions for allocating and freeing the skip lists. We will add them later, if
+ * we agree that this is the right overall approach.
+ */
+
+/*
+ * TMPL_FN_APPEND_SEARCH --
+ *     Fast append search of a skiplist, creating a skiplist stack as we go. In other words, we
+ * quickly check if the given item can be appended to the end of the skiplist, in which case we set
+ * the "done" flag to true, and create the stack accordingly. If the key we would like to insert is
+ * an exact match, return it via the "key" argument.
+ */
+static inline int
+TMPL_FN_APPEND_SEARCH(WT_SESSION_IMPL *session, TMPL_CURSOR *cbt, TMPL_HEAD *ins_head,
```

Comment:
Can we template the session as well? We can have a noop session for the testing code.

Reply:
I was thinking about removing the session argument, and I even found a nice workaround for the `collator` needing the session argument, but then I got to the point where we use locking—and that's when I decided it's probably not worth the effort at least at this stage.

We use locking whenever we need to update the tail of the skip list. (On a different note, I'd like to removing locking from here if I could, but it's hard, and it would result in performance tradeoffs elsewhere—so this is not yet the time to do it.)

Reply:
Thanks, gotcha.